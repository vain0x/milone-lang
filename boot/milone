#!/bin/bash

dotnet run -p MiloneLang -c Release "$@"
