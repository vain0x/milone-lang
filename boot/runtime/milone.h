// Runtime code for C programs generated by the milone-lang compiler.

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

struct String {
  char* str;
  int len;
};

int int_cmp(int l, int r) {
  if (l == r) return 0;
  if (l < r) return -1;
  return 1;
}

int int_clamp(int x, int l, int r) {
  if (x < l) return l;
  if (x > r) return r;
  return x;
}

int str_cmp(struct String left, struct String right) {
  return strcmp(left.str, right.str);
}

struct String str_add(struct String left, struct String right) {
  if (left.len == 0 || right.len == 0) {
    return right.len == 0 ? left : right;
  }
  int len = left.len + right.len;
  char* str = (char*)malloc((len + 1) * sizeof(char));
  strcpy(str, left.str);
  strcpy(str + left.len, right.str);
  return (struct String){.str = str, .len = len};
}

struct String str_substring(int l, int r, struct String s) {
  l = int_clamp(l, 0, s.len);
  r = int_clamp(r + 1, l, s.len);
  int len = r - l;
  char* str;
  if (r == s.len) {
    str = s.str + l;
  } else {
    str = (char*)malloc((len + 1) * sizeof(char));
    strncpy(str, s.str + l, len);
    str[len] = '\0';
  }
  return (struct String){.str = str, .len = len};
}

int str_to_int(struct String s) {
  return atoi(s.str);
}

struct String str_of_int(int value) {
  char* str = (char *)calloc(20, sizeof(char));
  sprintf(str, "%d", value);
  return (struct String){.str = str, .len = strlen(str)};
}

struct String str_of_char(char value) {
  char* str = (char *)calloc(2, sizeof(char));
  str[0] = value;
  return (struct String){.str = str, .len = strlen(str)};
}

void milone_assert(int cond) {
  if (!cond) {
    exit(1);
  }
}

struct String file_read_all_text(struct String file_name) {
  FILE *fp = fopen(file_name.str, "r");
  if (!fp) {
    fprintf(stderr, "File '%s' not found.", file_name.str);
    abort();
  }

  fseek(fp, 0, SEEK_END);
  long size = ftell(fp);
  if (size < 0) {
    fclose(fp);
    fprintf(stderr, "%s", "Couldn't retrieve the file size.");
    abort();
  }
  fseek(fp, 0, SEEK_SET);

  char *content = (char *)calloc((size_t)size + 1, sizeof(char));
  size_t read_size = fread(content, 1, (size_t)size, fp);
  if (read_size != (size_t)size) {
    fclose(fp);
    fprintf(stderr, "%s", "Couldn't retrieve the file contents");
    abort();
  }

  fclose(fp);
  return (struct String){.str = content, .len = size};
}

int file_write_all_text(struct String file_name, struct String content) {
  FILE *fp = fopen(file_name.str, "w");
  if (!fp) {
    fprintf(stderr, "File '%s' not found.", file_name.str);
    abort();
  }

  fprintf(fp, "%s", content.str);

  fclose(fp);
  return 0;
}

static int s_argc;
static char **s_argv;

int arg_count(int _unit) {
  return s_argc;
}

struct String arg_get(int index) {
  if (!(0 <= index && index < s_argc)) {
    abort();
  }

  char *str = s_argv[index];
  int len = strlen(str);
  return (struct String){.str = str, .len = len};
}

int main(int argc, char** argv) {
  s_argc = argc - 1;
  s_argv = argv + 1;

  int milone_main();
  return milone_main();
}

#define main milone_main
