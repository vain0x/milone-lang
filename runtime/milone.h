// Runtime code for C programs generated by the milone-lang compiler.
// Interface is still unstable.

#ifndef MILONE_H_INCLUDED
#define MILONE_H_INCLUDED

#include <stdbool.h>
#include <stddef.h>
#include <stdlib.h>
#include <stdint.h>
#include <stdio.h>

// -----------------------------------------------
// assertion
// -----------------------------------------------

void milone_assert(bool cond, int row, int column);

// -----------------------------------------------
// memory management
// -----------------------------------------------

void milone_enter_region(void);
void milone_leave_region(void);
void *milone_mem_alloc(int count, size_t size);

// -----------------------------------------------
// int
// -----------------------------------------------

int int_compare(int l, int r);
int int64_compare(int64_t l, int64_t r);
int uint64_compare(uint64_t l, uint64_t r);

// -----------------------------------------------
// str
// -----------------------------------------------

// This assumes some invariants, see documentation about string type.
struct String {
    char const *str;
    int len;
};

// Convert a null-terminated string to a string object by computing its length.
//
// SAFETY: The resulting string is valid until the pointer is valid.
struct String str_borrow(char const *c_str);

// Compare two strings in lexicographical order.
int str_compare(struct String l, struct String r);

// Create a copy of native C string.
struct String str_of_raw_parts(char const *ptr, int len);

// Create a copy of native null-terminated C string.
struct String str_of_c_str(char const *s);

// Create a concatenation of two string.
struct String str_add(struct String l, struct String r);

// Create a slice of string.
struct String str_get_slice(int l, int r, struct String s);

struct String str_ensure_null_terminated(struct String s);

// Ensure null-terminated.
char const *str_to_c_str(struct String s);

// Conversion:

int8_t str_to_int8(struct String s);
int16_t str_to_int16(struct String s);
int str_to_int(struct String s);
int64_t str_to_int64(struct String s);
intptr_t str_to_intptr(struct String s);

uint8_t str_to_uint8(struct String s);
uint16_t str_to_uint16(struct String s);
uint32_t str_to_uint32(struct String s);
uint64_t str_to_uint64(struct String s);
uintptr_t str_to_uintptr(struct String s);

struct String str_of_int64(int64_t value);
struct String str_of_uint64(uint64_t value);

double str_to_double(struct String s);
struct String str_of_double(double value);

char str_to_char(struct String s);
struct String str_of_char(char value);

// Actual name of string list.
struct StringList;

struct String str_concat(struct String sep, struct StringList const *strings);

// Low level operations.

inline char const *str_as_ptr(struct String s) { return s.str; }

#endif
