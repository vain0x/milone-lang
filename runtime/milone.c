// Runtime code for C programs generated by the milone-lang compiler.

// Some of functions are not declared in milone.h
// but usable with __nativeFun.

#include <assert.h>
#include <ctype.h>
#include <errno.h>
#include <stdbool.h>
#include <stddef.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#include <milone.h>

// -----------------------------------------------
// memory management (memory pool)
// -----------------------------------------------

// structure for memory management. poor implementation. thread unsafe.
struct MemoryChunk {
    // start of allocated memory or null
    void *ptr;

    // used size
    size_t len;

    // total capacity
    size_t cap;

    // list of used chunks or null
    // each chunk is allocated by calloc and should be freed after leaving
    // region
    struct MemoryChunk *used;

    // parent region or null
    // each region is allocated by calloc and should be freed after leaving
    // region
    struct MemoryChunk *parent;
};

static struct MemoryChunk s_heap;
static int s_heap_level; // depth of current region
static int s_heap_size;  // consumed size in all regions
static int s_heap_alloc; // allocated size in all regions

static void oom(void) {
    fprintf(stderr, "Out of memory.\n");
    exit(1);
}

// free chunk itself, not used list
static void free_chunk(struct MemoryChunk *chunk) {
    assert(chunk != NULL);
    assert(chunk->ptr != NULL);

    free(chunk->ptr);
    s_heap_size -= chunk->len;
    s_heap_alloc -= chunk->cap;
}

void milone_enter_region(void) {
    // fprintf(stderr, "debug: enter_region level=%d size=%d\n", s_heap_level +
    // 1, s_heap_size);

    struct MemoryChunk *parent = calloc(1, sizeof(struct MemoryChunk));
    if (parent == NULL) {
        oom();
    }

    *parent = s_heap;
    s_heap = (struct MemoryChunk){.parent = parent};

    s_heap_level++;
}

void milone_leave_region(void) {
    // fprintf(stderr, "debug: leave_region level=%d size=%d\n", s_heap_level,
    // s_heap_size);

    assert(s_heap.parent != NULL);
    struct MemoryChunk *parent = s_heap.parent;

    // free current region
    {
        if (s_heap.ptr != NULL) {
            free_chunk(&s_heap);
        }

        // other than s_heap, chunk itself needs to be freed
        struct MemoryChunk *chunk = s_heap.used;
        while (chunk) {
            struct MemoryChunk *current = chunk;
            chunk = current->used;

            free_chunk(current);
            free(current);
        }

        s_heap = *parent;
        free(parent);
    }

    assert(s_heap_level > 0);
    s_heap_level--;

    // fprintf(stderr, "debug: free level=%d size=%d\n", s_heap_level,
    // s_heap_size);
}

// allocate new chunk
static void *milone_mem_alloc_slow(size_t total) {
    size_t cap = s_heap.cap;

    // move current chunk to used list
    if (s_heap.ptr != NULL) {
        struct MemoryChunk *used = calloc(1, sizeof(struct MemoryChunk));
        if (used == NULL) {
            oom();
        }

        *used = s_heap;
        s_heap.used = used;
    }

    // grow exponentially
    cap *= 2;
    if (cap < total) {
        cap = total;
    }

    // initialize chunk
    void *ptr = calloc(1, cap);
    if (ptr == NULL) {
        oom();
    }
    s_heap.ptr = ptr;
    s_heap.len = total;
    s_heap.cap = cap;

    s_heap_size += total;
    s_heap_alloc += cap;

    // fprintf(stderr, "debug: alloc level=%d size=%d  new chunk(h=%d
    // cap=%d)\n", s_heap_level, (int)size, h, (int)cap);
    return ptr;
}

void *milone_mem_alloc(int count, size_t size) {
    assert(count > 0 && size > 0);

    size_t total = (size_t)count * size;

    // align to 16x
    total = ((total & ~15) | 15) + 1;
    assert(total % 16 == 0);

    if (s_heap.len + total > s_heap.cap) {
        return milone_mem_alloc_slow(total);
    }

    // use current chunk
    void *ptr = (char *)s_heap.ptr + s_heap.len;
    s_heap.len += total;
    s_heap_size += total;

    // fprintf(stderr, "debug: alloc level=%d size=%d (%dx%d)\n", s_heap_level,
    // (int)((size_t)count * size), (int)count, (int)size);
    return ptr;
}

// -----------------------------------------------
// int
// -----------------------------------------------

int int_compare(int l, int r) {
    if (l == r)
        return 0;
    if (l < r)
        return -1;
    return 1;
}

int int64_compare(int64_t l, int64_t r) {
    if (l == r)
        return 0;
    if (l < r)
        return -1;
    return 1;
}

int uint64_compare(uint64_t l, uint64_t r) {
    if (l == r)
        return 0;
    if (l < r)
        return -1;
    return 1;
}

int int_clamp(int x, int l, int r) {
    if (x < l)
        return l;
    if (x > r)
        return r;
    return x;
}

// -----------------------------------------------
// str
// -----------------------------------------------

struct StringBuilder {
    char *buf;
    int len;
    int cap;
};

static struct StringBuilder *string_builder_new_with_capacity(int cap) {
    assert(cap >= 0);

    struct StringBuilder *sb =
        milone_mem_alloc(1, sizeof(struct StringBuilder));
    *sb = (struct StringBuilder){
        .buf = cap > 0 ? milone_mem_alloc(cap, sizeof(char)) : "",
        .len = 0,
        .cap = cap,
    };
    return sb;
}

static void string_builder_grow(struct StringBuilder *sb, int addition) {
    assert(addition >= 0);
    assert(sb->len + addition >= sb->cap);

    // grow exponentially
    sb->cap *= 2;

    int min_len = sb->len + addition + 1;
    if (sb->cap < min_len) {
        sb->cap = min_len;
    }

    char *buf = milone_mem_alloc(sb->cap, sizeof(char));
    memcpy(buf, sb->buf, sb->len);
    sb->buf = buf;

    assert(sb->len < sb->cap);
    sb->buf[sb->len] = '\0';
}

static void string_builder_append_string(struct StringBuilder *sb,
                                         struct String value) {
    if (sb->len + value.len >= sb->cap) {
        string_builder_grow(sb, value.len);
    }

    memcpy(&sb->buf[sb->len], value.str, value.len + 1);
    sb->len += value.len;
    assert(sb->buf[sb->len] == '\0');
}

int str_cmp(struct String left, struct String right) {
    int min_len = int_clamp(left.len, 0, right.len);
    return memcmp(left.str, right.str, min_len + 1);
}

/// Create a string from a slice of native C string.
struct String str_of_raw_parts(char const *p, int len) {
    if (len <= 0) {
        if (len < 0) {
            fprintf(stderr, "FATAL: Negative string length (%d).\n", len);
            abort();
        }

        return (struct String){.str = "", .len = 0};
    }

    char *str = (char *)milone_mem_alloc(len + 1, sizeof(char));
    memcpy(str, p, len * sizeof(char));
    str[len] = '\0';
    return (struct String){.str = str, .len = len};
}

struct String str_add(struct String left, struct String right) {
    if (left.len == 0 || right.len == 0) {
        return right.len == 0 ? left : right;
    }
    int len = left.len + right.len;
    char *str = (char *)milone_mem_alloc(len + 1, sizeof(char));
    memcpy(str, left.str, left.len);
    memcpy(str + left.len, right.str, right.len);
    assert(str[len] == '\0');
    return (struct String){.str = str, .len = len};
}

struct String str_get_slice(int l, int r, struct String s) {
    l = int_clamp(l, 0, s.len);
    r = int_clamp(r + 1, l, s.len);
    int len = r - l;
    char *str;
    if (r == s.len) {
        str = s.str + l;
    } else {
        str = (char *)milone_mem_alloc(len + 1, sizeof(char));
        memcpy(str, s.str + l, len);
    }
    assert(str[len] == '\0');
    return (struct String){.str = str, .len = len};
}

static void verify_str_to_number(const char *type_name, const char *endptr,
                                 bool range_check) {
    if (!range_check || (*endptr != '\0' && !isspace(*endptr)) ||
        errno == ERANGE) {
        fprintf(stderr, "FATAL: Failed to convert a string to %s.\n",
                type_name);
        exit(1);
    }
}

int8_t str_to_int8(struct String s) {
    char *endptr = s.str + s.len;
    int n = strtol(s.str, &endptr, 10);
    verify_str_to_number("int8_t", endptr, INT8_MIN <= n && n <= INT8_MAX);
    return (int8_t)n;
}

int16_t str_to_int16(struct String s) {
    char *endptr = s.str + s.len;
    int n = strtol(s.str, &endptr, 10);
    verify_str_to_number("int16_t", endptr, INT16_MIN <= n && n <= INT16_MAX);
    return (int16_t)n;
}

int str_to_int(struct String s) {
    char *endptr = s.str + s.len;
    int n = strtol(s.str, &endptr, 10);
    verify_str_to_number("int", endptr, true);
    return n;
}

int64_t str_to_int64(struct String s) {
    char *endptr = s.str + s.len;
    int64_t n = strtoll(s.str, &endptr, 10);
    verify_str_to_number("int64_t", endptr, true);
    return n;
}

intptr_t str_to_intptr(struct String s) {
    char *endptr = s.str + s.len;
    int64_t n = strtoll(s.str, &endptr, 10);
    verify_str_to_number("intptr_t", endptr, true);
    return (intptr_t)n;
}

uint8_t str_to_uint8(struct String s) {
    char *endptr = s.str + s.len;
    uint32_t n = strtoul(s.str, &endptr, 10);
    verify_str_to_number("uint8_t", endptr, n <= UINT8_MAX);
    return (uint8_t)n;
}

uint16_t str_to_uint16(struct String s) {
    char *endptr = s.str + s.len;
    uint32_t n = strtoul(s.str, &endptr, 10);
    verify_str_to_number("uint16_t", endptr, n <= UINT16_MAX);
    return (uint16_t)n;
}

uint32_t str_to_uint32(struct String s) {
    char *endptr = s.str + s.len;
    uint32_t n = strtoul(s.str, &endptr, 10);
    verify_str_to_number("uint32_t", endptr, true);
    return n;
}

uint64_t str_to_uint64(struct String s) {
    char *endptr = s.str + s.len;
    uint64_t n = strtoull(s.str, &endptr, 10);
    verify_str_to_number("uint64_t", endptr, true);
    return n;
}

uintptr_t str_to_uintptr(struct String s) {
    char *endptr = s.str + s.len;
    uint64_t n = strtoull(s.str, &endptr, 10);
    verify_str_to_number("uintptr_t", endptr, true);
    return n;
}

struct String str_of_int64(int64_t value) {
    char buf[21] = {};
    int n = sprintf(buf, "%ld", value);
    return str_of_raw_parts(buf, n);
}

struct String str_of_uint64(uint64_t value) {
    char buf[21] = {};
    int n = sprintf(buf, "%lu", value);
    return str_of_raw_parts(buf, n);
}

double str_to_double(struct String s) {
    char *endptr = s.str + s.len;
    double n = strtod(s.str, &endptr);
    verify_str_to_number("float", endptr, true);
    return n;
}

struct String str_of_double(double value) {
    char buf[64] = {};
    int n = sprintf(buf, "%f", value);
    return str_of_raw_parts(buf, n);
}

struct String str_of_char(char value) {
    char *str = (char *)milone_mem_alloc(2, sizeof(char));
    str[0] = value;
    return (struct String){.str = str, .len = strlen(str)};
}

struct MyStringList {
    struct String head;
    struct MyStringList *tail;
};

struct String str_concat(struct String sep, struct StringList *strings) {
    struct MyStringList *ss = (struct MyStringList *)strings;

    struct StringBuilder *sb = string_builder_new_with_capacity(0x1000);
    bool first = true;

    while (ss) {
        struct String head = ss->head;
        ss = ss->tail;

        if (!first && sep.len > 0) {
            string_builder_append_string(sb, sep);
        }
        first = false;

        string_builder_append_string(sb, head);
    }

    return str_of_raw_parts(sb->buf, sb->len);
}

// -----------------------------------------------
// assertion
// -----------------------------------------------

void milone_assert(bool cond, int y, int x) {
    if (!cond) {
        fprintf(stderr, "Assertion failed at (%d, %d)\n", y + 1, x + 1);
        exit(1);
    }
}

// -----------------------------------------------
// file IO
// -----------------------------------------------

int file_exists(struct String file_name) {
    bool ok = false;

    FILE *fp = fopen(file_name.str, "r");
    if (fp) {
        ok = true;
        fclose(fp);
    }

    return ok;
}

struct String file_read_all_text(struct String file_name) {
    FILE *fp = fopen(file_name.str, "r");
    if (!fp) {
        fprintf(stderr, "File '%s' not found.", file_name.str);
        abort();
    }

    fseek(fp, 0, SEEK_END);
    long size = ftell(fp);
    if (size < 0) {
        fclose(fp);
        fprintf(stderr, "%s", "Couldn't retrieve the file size.");
        abort();
    }
    fseek(fp, 0, SEEK_SET);

    char *content = (char *)milone_mem_alloc((size_t)size + 1, sizeof(char));
    size_t read_size = fread(content, 1, (size_t)size, fp);
    if (read_size != (size_t)size) {
        fclose(fp);
        fprintf(stderr, "%s", "Couldn't retrieve the file contents");
        abort();
    }

    fclose(fp);
    return (struct String){.str = content, .len = size};
}

void file_write_all_text(struct String file_name, struct String content) {
    FILE *fp = fopen(file_name.str, "w");
    if (!fp) {
        fprintf(stderr, "File '%s' not found.", file_name.str);
        abort();
    }

    fprintf(fp, "%s", content.str);

    fclose(fp);
}

struct String milone_get_env(struct String name) {
    char const *value = getenv(name.str);
    if (value == NULL) {
        return (struct String){.str = "", .len = 0};
    }

    return str_of_raw_parts(value, strlen(value));
}

// -----------------------------------------------
// profiling
// -----------------------------------------------

static long milone_get_time_millis(void) {
    struct timespec t;
    timespec_get(&t, TIME_UTC);
    return t.tv_sec * 1000L + t.tv_nsec / (1000L * 1000L);
}

struct Profiler {
    long epoch;
    long heap_size;
};

void *milone_profile_init(void) {
    struct Profiler *p =
        (struct Profiler *)milone_mem_alloc(1, sizeof(struct Profiler));
    p->epoch = milone_get_time_millis();
    p->heap_size = s_heap_size;
    return p;
}

void milone_profile_log(struct String msg, void *profiler) {
    struct Profiler *p = (struct Profiler *)profiler;

    long t = milone_get_time_millis();
    long s = p->epoch;

    long millis = t - s;
    if (millis < 0) {
        millis = 0;
    }

    long sec = millis / 1000;
    millis %= 1000;

    long bytes = s_heap_size - p->heap_size;
    if (bytes < 0) {
        bytes = 0;
    }

    long kilo = bytes / 1000;
    bytes %= 1000;

    long mega = kilo / 1000;
    kilo %= 1000;

    double usage = s_heap_alloc == 0 ? 0.0 : (double)s_heap_size / s_heap_alloc;
    fprintf(stderr, "profile: time=%4d.%03d mem=%5d,%03d,%03d use=%.03f\n%s\n",
            (int)sec, (int)millis, (int)mega, (int)kilo, (int)bytes, usage,
            msg.str);

    p->epoch = t;
    p->heap_size = s_heap_size;
}

// -----------------------------------------------
// For competitive programming
// -----------------------------------------------

int scan_int(void) {
    int value;
    int _n = scanf("%d", &value);
    return value;
}

char scan_char(void) {
    char value;
    int _n = scanf("%c", &value);
    return value;
}

struct String scan_str(int capacity) {
    if (capacity <= 0) {
        fprintf(stderr, "scan_str(%d)", capacity);
        exit(1);
    }

    char *str = calloc(capacity, sizeof(char));
    if (str == NULL) {
        fprintf(stderr, "scan_str(%d) out of memory", capacity);
        exit(1);
    }

    char fmt[16] = {};
    sprintf(fmt, "%%%ds", capacity);
    assert(fmt[15] == 0);

    int _n = scanf(fmt, str);

    int len = strlen(str);
    assert(len < capacity);
    return (struct String){.str = str, .len = len};
}

void *int_array_new(int len) {
    void *p = calloc(len, sizeof(int));
    // fprintf(stderr, "int_array_new(len=%d) = %p\n", len, p);
    return p;
}

int int_array_get(void *array, int index) {
    int value = ((int *)array)[index];
    // fprintf(stderr, "int_array_get(%p, index=%d) = %d\n", array, index,
    // value);
    return value;
}

void int_array_set(void *array, int index, int value) {
    // fprintf(stderr, "int_array_set(%p, index=%d, value=%d)\n", array, index,
    // value);
    ((int *)array)[index] = value;
}

void int_array_copy(void *dest, void *src, int size) {
    memcpy(dest, src, size);
}

// -----------------------------------------------
// Command-line Arguments
// -----------------------------------------------

static int s_argc;
static char **s_argv;

int arg_count(void) { return s_argc; }

struct String arg_get(int index) {
    if (!(0 <= index && index < s_argc)) {
        abort();
    }

    char *str = s_argv[index];
    int len = strlen(str);
    return (struct String){.str = str, .len = len};
}

// -----------------------------------------------
// Runtime Entrypoint
// -----------------------------------------------

#undef main

int main(int argc, char **argv) {
    s_argc = argc - 1;
    s_argv = argv + 1;

    int milone_main();
    return milone_main();
}
