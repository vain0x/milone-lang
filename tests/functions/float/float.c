#include "milone.h"

struct F64BoolFun1;

struct F64List;

int go_1(struct F64BoolFun1 pred_, struct F64List* xs_1);

int all_1(struct F64BoolFun1 pred_, struct F64List* xs_);

int fun_(double x_1);

int fun_2(void* env_, double arg_);

int fun_1(double x_2);

int fun_3(void* env_1, double arg_1);

int main();

struct F64BoolFun1 {
    int(*fun)(void*, double);
    void* env;
};

struct F64List {
    double head;
    struct F64List* tail;
};

int go_1(struct F64BoolFun1 pred_, struct F64List* xs_1) {
tailrec_1:;
    int match_;
    if ((!((!(xs_1))))) goto next_3;
    match_ = 1;
    goto end_match_2;
next_3:;
    if ((!(xs_1))) goto next_4;
    double x_ = xs_1->head;
    struct F64List* xs_2 = xs_1->tail;
    int app_ = pred_.fun(pred_.env, x_);
    int if_;
    if (app_) {
        goto then_6;
    } else {
        goto else_7;
    }
then_6:;
    struct F64BoolFun1 arg_2 = pred_;
    struct F64List* arg_3 = xs_2;
    pred_ = arg_2;
    xs_1 = arg_3;
    goto tailrec_1;
    if_ = 0;
    goto if_next_5;
else_7:;
    if_ = 0;
    goto if_next_5;
if_next_5:;
    match_ = if_;
    goto end_match_2;
next_4:;
    exit(1);
end_match_2:;
    return match_;
}

int all_1(struct F64BoolFun1 pred_, struct F64List* xs_) {
    int call_ = go_1(pred_, xs_);
    return call_;
}

int fun_(double x_1) {
    return (x_1 == 0.0);
}

int fun_2(void* env_, double arg_) {
    int call_1 = fun_(arg_);
    return call_1;
}

int fun_1(double x_2) {
    return (0.0 < x_2);
}

int fun_3(void* env_1, double arg_1) {
    int call_2 = fun_1(arg_1);
    return call_2;
}

int main() {
    struct F64List* list_7 = (struct F64List*)milone_mem_alloc(1, sizeof(struct F64List));
    list_7->head = 00.00e-00;
    list_7->tail = NULL;
    struct F64List* list_6 = (struct F64List*)milone_mem_alloc(1, sizeof(struct F64List));
    list_6->head = 0E-9;
    list_6->tail = list_7;
    struct F64List* list_5 = (struct F64List*)milone_mem_alloc(1, sizeof(struct F64List));
    list_5->head = 0E+9;
    list_5->tail = list_6;
    struct F64List* list_4 = (struct F64List*)milone_mem_alloc(1, sizeof(struct F64List));
    list_4->head = 0E0;
    list_4->tail = list_5;
    struct F64List* list_3 = (struct F64List*)milone_mem_alloc(1, sizeof(struct F64List));
    list_3->head = 0e0;
    list_3->tail = list_4;
    struct F64List* list_2 = (struct F64List*)milone_mem_alloc(1, sizeof(struct F64List));
    list_2->head = 000.000;
    list_2->tail = list_3;
    struct F64List* list_1 = (struct F64List*)milone_mem_alloc(1, sizeof(struct F64List));
    list_1->head = 0.;
    list_1->tail = list_2;
    struct F64List* list_ = (struct F64List*)milone_mem_alloc(1, sizeof(struct F64List));
    list_->head = 0.0;
    list_->tail = list_1;
    struct F64List* zeros_ = list_;
    void* box_ = (void*)milone_mem_alloc(1, sizeof(int));
    (*(((int*)box_))) = 0;
    void* env_2 = box_;
    struct F64BoolFun1 fun_4 = (struct F64BoolFun1){.fun = fun_2, .env = env_2};
    int call_3 = all_1(fun_4, zeros_);
    milone_assert(call_3, 24, 2);
    struct F64List* list_18 = (struct F64List*)milone_mem_alloc(1, sizeof(struct F64List));
    list_18->head = 6.02e-23;
    list_18->tail = NULL;
    struct F64List* list_17 = (struct F64List*)milone_mem_alloc(1, sizeof(struct F64List));
    list_17->head = 1e-23;
    list_17->tail = list_18;
    struct F64List* list_16 = (struct F64List*)milone_mem_alloc(1, sizeof(struct F64List));
    list_16->head = 1e+23;
    list_16->tail = list_17;
    struct F64List* list_15 = (struct F64List*)milone_mem_alloc(1, sizeof(struct F64List));
    list_15->head = 1E-9;
    list_15->tail = list_16;
    struct F64List* list_14 = (struct F64List*)milone_mem_alloc(1, sizeof(struct F64List));
    list_14->head = 1e-9;
    list_14->tail = list_15;
    struct F64List* list_13 = (struct F64List*)milone_mem_alloc(1, sizeof(struct F64List));
    list_13->head = 1E+9;
    list_13->tail = list_14;
    struct F64List* list_12 = (struct F64List*)milone_mem_alloc(1, sizeof(struct F64List));
    list_12->head = 1e+9;
    list_12->tail = list_13;
    struct F64List* list_11 = (struct F64List*)milone_mem_alloc(1, sizeof(struct F64List));
    list_11->head = 1E9;
    list_11->tail = list_12;
    struct F64List* list_10 = (struct F64List*)milone_mem_alloc(1, sizeof(struct F64List));
    list_10->head = 1e9;
    list_10->tail = list_11;
    struct F64List* list_9 = (struct F64List*)milone_mem_alloc(1, sizeof(struct F64List));
    list_9->head = 42.195;
    list_9->tail = list_10;
    struct F64List* list_8 = (struct F64List*)milone_mem_alloc(1, sizeof(struct F64List));
    list_8->head = 1.0;
    list_8->tail = list_9;
    struct F64List* values_ = list_8;
    void* box_1 = (void*)milone_mem_alloc(1, sizeof(int));
    (*(((int*)box_1))) = 0;
    void* env_3 = box_1;
    struct F64BoolFun1 fun_5 = (struct F64BoolFun1){.fun = fun_3, .env = env_3};
    int call_4 = all_1(fun_5, values_);
    milone_assert(call_4, 39, 2);
    milone_assert((6.0 < (2.3 + 4.5)), 42, 2);
    milone_assert(((44.0 - 2.0) == 42.0), 43, 2);
    milone_assert(((2.25 * 2.0) == 4.5), 44, 2);
    milone_assert(((8.0 / 2.0) == 4.0), 45, 2);
    milone_assert((0.0 == 0.0), 48, 2);
    milone_assert((0.0 != 1.0), 49, 2);
    milone_assert((2.0 < 3.0), 51, 2);
    milone_assert((2.0 >= 2.0), 52, 2);
    milone_assert((3.0 >= 2.0), 53, 2);
    milone_assert((2.0 < 3.0), 55, 2);
    milone_assert((2.0 >= 2.0), 56, 2);
    milone_assert((3.0 >= 2.0), 57, 2);
    milone_assert((((int)3.14) == 3), 60, 2);
    milone_assert((((double)3) == 3.0), 61, 2);
    struct String call_5 = str_of_double(3.14);
    milone_assert((str_cmp(call_5, (struct String){.str = "3.140000", .len = 8}) == 0), 65, 2);
    double call_6 = str_to_double((struct String){.str = "3.14", .len = 4});
    milone_assert((call_6 == 3.14), 68, 2);
    printf("Assume PI is %f.\n", 3.14);
    return 0;
}
