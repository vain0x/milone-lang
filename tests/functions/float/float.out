Assume PI is 3.140000.
$? = 0
